presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClient.java

```diff
@@ -346,39 +366,82 @@ public String quoted(RemoteTableName remoteTableName)
         return delegate.getTableScanRedirection(session, tableHandle);
     }
 
+    private ColumnsCacheKey buildColumnsCacheKey(ConnectorSession session, SchemaTableName tableName)
+    {
+        return new ColumnsCacheKey(JdbcIdentity.from(session), getSessionProperties(session), tableName);
+    }
+
+    private TableHandleCacheKey buildTableHandleCacheKey(ConnectorSession session, SchemaTableName schemaTableName)
+    {
+        return new TableHandleCacheKey(JdbcIdentity.from(session), getSessionProperties(session), schemaTableName);
+    }
+
+    private TableNamesCacheKey buildTableNamesCacheKey(ConnectorSession session, Optional<String> schema)
+    {
+        return new TableNamesCacheKey(JdbcIdentity.from(session), getSessionProperties(session), schema);
+    }
+
     private void invalidateSchemasCache()
     {
         schemaNamesCache.invalidateAll();
     }
 
-    private void invalidateTablesCaches()
+    private void invalidateTablesCaches(SchemaTableName table)
+    {
+        invalidateColumnsCache(table);
+        invalidateTableCache(table);
+    }
+
+    private void invalidateTableCache(SchemaTableName table)
+    {
+        invalidateCache(tableNamesCache, key -> key.schemaName.equals(Optional.of(table.getSchemaName())));
```

Comment:
what when `key.schemaName.isEmpty()`? How that can happen?

Reply:
nothing will be invalidated